[stdlib] Switch to a stable sort algorithm

natecook1000 opened the pull request:
This switches the standard library's sort algorithm from an in-place introsort to use a modified timsort, a stable, adaptive sort that merges runs using a temporary buffer. This implementation performs straight merges instead of adopting timsort's galloping strategy.

In addition to maintaining the relative order of equal/non-comparable elements, this algorithm outperforms the introsort on data with any intrinsic structure, such as runs of ascending or descending elements or a significant number of equality collisions. Benchmarking script and results can be seen in this gist: https://gist.github.com/natecook1000/5161e10aeba09408c130284ea6ec4e11

natecook1000 committed:
[stdlib] Add stable sort to mutable collections

This switches the standard library's sort algorithm from an in-place
introsort to use a modified timsort, a stable, adaptive sort that
merges runs using a temporary buffer. This implementation performs
straight merges instead of adopting timsort's galloping strategy.

In addition to maintaining the relative order of equal/non-comparable
elements, this algorithm outperforms the introsort on data with any
intrinsic structure, such as runs of ascending or descending elements
or a significant number of equality collisions.

natecook1000 committed:
[stdlib] Move the old sort algorithm into a test

natecook1000 committed:
[test] Update the `sort` test to check for stability

natecook1000 committed:
[stdlib] Revise `sort` and `sorted` docs to note stability change.

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented:
@swift-ci Please test

airspeedswift commented on stdlib/public/core/Sort.swift:
```diff
@@ -146,9 +146,6 @@ where Self: RandomAccessCollection, Element: Comparable {
   /// `Comparable` protocol by calling this method. Elements are sorted in
   /// ascending order.
   ///
-  /// The sorting algorithm is not stable. A nonstable sort may change the
```
I'd suggest leaving this in for now. If this sort is better than the current one, we can just land it as an implementation improvement. But officially committing to stability needs an evolution proposal.

natecook1000 replied:
Ah, good point.

xwu replied:
Well, it might be misleading to say it's "not stable" when it is: I'd suggest maybe wording it something like "is not guaranteed to be stable" or "may or may not be stable," or maybe explicitly, "It is an implementation detail whether the sorting algorithm is stable."

airspeedswift replied:
Yeah, "not guaranteed stable" is the right phrasing.

airspeedswift commented:
@swift-ci please test compiler performance

natecook1000 committed:
Back off the sort stability claims

natecook1000 committed:
Switch precondition to _sanityCheck

natecook1000 committed:
Work around possible optimizer issue

See rdar://45044610

natecook1000 commented:
@swift-ci Please smoke benchmark

natecook1000 commented:
@swift-ci Please smoke benchmark

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
-        // Move y forward
-        self[i] = predecessor
+        
+        // Swap the elements at `i` and `j`.
```
Nice comment =)

natecook1000 replied:
😅 I think I was just excited about not using the temporary any more

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
-      // 1 swap: 132, 121 --- 2 swaps: 231, 221
-      // swap(b, c): 132->123, 121->112, 231->213, 221->212
-      swapAt(b, c)
+  return true
```
Doesn't look like there are any other `return`s here. So perhaps the whole `-> Bool` is not needed?

natecook1000 replied:
Definitely not needed — it's a workaround for an optimizer issue. Those can all come out once that's resolved.

airspeedswift replied:
optimizer issues also suggest the benchmarks could improve once resolved...

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
+  if c < 1 << bitsToUse {
+    return c
+  }
+  let offset = (MemoryLayout<Int>.size * 8 - bitsToUse) - c.leadingZeroBitCount
```
Isn't `MemoryLayout<Int>.size * 8` equal to `Int.bitWidth`?

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
+  if next < elements.endIndex {
+    var current = next
+    elements.formIndex(after: &next)
+    if try next < elements.endIndex &&
```
Can we extract this `next < elements.endIndex` as a guard, and then replace the `else if` few lines below with just an `else`?

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
+    } else if next < elements.endIndex {
+      // The elements in this run are in ascending order. Equal elements can
+      // be included in the run.
+      repeat {
```
This `repeat ... while` duplication is annoying. Would extracting it into a nested function affect performance?

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
+      assert(runs.count == 1, "Didn't complete final merge")
     }
+
+    if !result { fatalError() }
```
Isn't it the same as `precondition(result)`?

moiseev commented on test/Prototypes/IntroSort.swift:
```diff
@@ -0,0 +1,397 @@
+// RUN: %target-run %t/a.out
+// REQUIRES: executable_test
+
+#if USE_STDLIBUNITTEST
```
Is this really needed? If so, the same guard should wrap around the actual test cases and a call to `runAllTests`.

moiseev commented on test/Prototypes/IntroSort.swift:
```diff
@@ -0,0 +1,397 @@
+// RUN: %target-codesign %t/a.out
+// RUN: %target-run %t/a.out
+// REQUIRES: executable_test
+
```
Maybe a "historical note" would be in order. As in, "this used to be the standard library preferred sort algorithm until Swift 5.0" or something like that.

moiseev commented on validation-test/stdlib/Sort.swift.gyb:
```diff
@@ -56,6 +56,12 @@ func expectSortedCollection(_ sortedAry: ArraySlice<Int>, _ originalAry: ArraySl
+func expectSortedCollection<C: Collection>(_ a: C,
+  by areInIncreasingOrder: (C.Element, C.Element) -> Bool
+) {
+  expectFalse(zip(a.dropFirst(), a).contains(where: areInIncreasingOrder))
```
Of course you meant `a.lazy.dropFirst()` :-)

moiseev replied:
I don't think so. It's just that I seem to have developed this habit of trying to apply `.lazy` everywhere and see if that'd work. Works in this case. Will perhaps make the test a millisecond faster.

moiseev commented:
This is great, @natecook1000! 👏 👏 👏

natecook1000 committed:
Address Max's feedback

moiseev commented on stdlib/public/core/Sort.swift:
```diff
@@ -529,6 +528,8 @@ extension UnsafeMutableBufferPointer {
     runs[i - 1] = low..<high
     runs.remove(at: i)
+
+    // FIXME: Remove this, it works around rdar://problem/45044610
```
👍

moiseev reviewed (approved):
LGTM

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Update source stability test

natecook1000 commented:
@swift-ci Please smoke test Linux platform

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Switch to use _withUMBPIfSupported

This drops the new MutableCollection requirement in favor of using only
_withUnsafeMutableBufferPointerIfSupported to dispatch the sort algorithm
to a mutable buffer. This in turn requires that UnsafeMutableBufferPointer
implements _withUnsafeMutableBufferPointerIfSupported, so that change is also
included here (and makes more intuitive sense than a buffer not supporting
buffer operations).

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Merge branch 'master' into nc-stable-sort

natecook1000 committed:
Merge branch 'master' into nc-stable-sort

natecook1000 committed:
Test unsafe buffer partition and sort

natecook1000 commented:
@swift-ci Please smoke test

airspeedswift commented:
:shipit:

natecook1000 committed:
Merge branch 'master' into nc-stable-sort

natecook1000 committed:
Update source stability test

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Merge branch 'master' into nc-stable-sort

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test Linux platform

natecook1000 commented:
@swift-ci Please smoke test Linux platform

natecook1000 commented:
@swift-ci Please smoke test Linux platform

CemTitor reviewed (commented):
nice pr
